### PR TITLE
Streamline publishing process

### PR DESCRIPTION
**Pull request by bisgaard-itis**
## What do these changes do?
- Ensure client version comes from a `VERSION` file in `clients/python/client`.
- Add a make recipe which will generate a `VERSION` file. The version generated has the format `<latest released version>+<number of commits from latest released commits>`. This ensures every new pushed commit can be published on TestPypi with a unique version number. I have tested that `packaging.version.Version` behaves properly with respect to this type of versioning. E.g. `Version("3.2.6+15")>Version("3.2.6+14")` evaluates to `True` while `Version("3.2.6+15")>Version("3.2.7+14")` evaluates to `False`.
- Update the github workflow so it generates a correct `VERSION` file if the workflow is triggered by a tagged commit (=release) and every new commits on master is published to TestPypi.
- Clean up release workflow. Now releasing a client is done entirely through Github. Simply tag a commit with a tag of the type `v3.2.4`.

### CODE REVIEW

File: .github/workflows/publish-python-client.yml [hunk 1]
```diff
@@ -13,10 +13,16 @@ jobs:
...
           ref: ${{ github.event.client_payload.ref }}
+          fetch-depth: 0
       - name: Set up Python
         uses: actions/setup-python@v2
         with:
           python-version: 3.9
+      - name: Generate version from tag
+        if: github.event_name == 'push' && startsWith(github.ref, 'refs/tags')
+        run: |
+          tag_version=$(echo ${{ github.ref }} | grep -oP '(?<=refs/tags/v)\d+\.\d+\.\d+'))
```
Comment (wvangeit) [inline .github/workflows/publish-python-client.yml:24]: I think it's ok, but this will fail if somebody e.g. uses new version '3.0', right? It might be slightly unexpected, but the requirement could be imposed I guess.

**Review by wvangeit — APPROVED** [review]

File: .github/workflows/publish-python-client.yml [hunk 2]
```diff
@@ -162,12 +174,14 @@ jobs:
...
         run: rm -f dist/${{needs.build-n-publish.outputs.osparc_client}}
       - name: Publish osparc to Test PyPI
+        if: github.ref == 'refs/heads/master' && github.event_name == 'push'
         uses: pypa/gh-action-pypi-publish@release/v1
         with:
           repository-url: https://test.pypi.org/legacy/
           verbose: true
           packages-dir: dist/
       - name: Publish osparc to PyPI
+        if: github.event_name == 'push' && startsWith(github.ref, 'refs/tags')
```
Comment (wvangeit) [inline .github/workflows/publish-python-client.yml:184]: Actually, is this protected anywhere that it will only run on master?

**Review by wvangeit — CHANGES_REQUESTED** [review]

File: .github/workflows/publish-python-client.yml (hunk 2 above)
Comment (bisgaard-itis) [reply .github/workflows/publish-python-client.yml:184]: I was looking into how this can be checked but I couldn't see how to easily do it. Tbh I didn't spend a lot of time checking it, but the point is that the `github.ref` variable which is available here does not tell if this tagged commit comes from master. Perhaps I could check above if the current `HEAD` is the latest commit on `master`. Or do you have any better ideas? What I did here is the suggested way in the docu for the `gh-action-pypi-publishing` packaging we are using here.

File: .github/workflows/publish-python-client.yml (hunk 1 above)
Comment (bisgaard-itis) [reply .github/workflows/publish-python-client.yml:24]: I added a validation step, that a valid semantic version string is extracted: https://github.com/ITISFoundation/osparc-simcore-clients/pull/83/commits/18e9776b98c235f190cb3f988d9d8288e404b5e6

File: .github/workflows/publish-python-client.yml (hunk 2 above)
Comment (wvangeit) [reply .github/workflows/publish-python-client.yml:184]: Do you have a link to that docu? 
Maybe they only do it in an workflow that anyway only runs on master? (this file isn't as far as I can see?)

Comment (bisgaard-itis) [reply .github/workflows/publish-python-client.yml:184]: Yes, sorry it is not actually that documentation, but a link which is in that documentation: https://packaging.python.org/en/latest/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows/#defining-a-workflow-job-environment
I will investigate if there is an easy way to check this. But I think another point is that we should always be using forks for developing, so basically there should only be a single branch (=`master`) on the main repo. Not sure if that is currently the case though...

Comment (wvangeit) [reply .github/workflows/publish-python-client.yml:184]: I think on option is to protect the entire (or at least the relevant part of the file) file from only running on master:
https://github.com/BlueBrain/BluePyOpt/blob/5232a8f2e278cf0d32fe7008a81d82b0ad10b992/.github/workflows/build.yml#L3

Comment (wvangeit) [reply .github/workflows/publish-python-client.yml:184]: In my opinion, but I could be wrong, is to protect the entire file with:
`on:
  push:
    branches: [master]
`
and then protect the specific task with the startsWith(github.ref, 'refs/tags')

Comment (wvangeit) [reply .github/workflows/publish-python-client.yml:184]: Actually, also suggested here: https://stackoverflow.com/a/66463380

Comment (bisgaard-itis) [reply .github/workflows/publish-python-client.yml:184]: Thanks. The thing is that, in the current setup, this workflow should also be run on PRs, so not only the master branch. Maybe I need to split the workflow into two workflows. I can try out what you suggest, but I am a bit skeptical about what they suggest in the stackoverflow link. As far as I can tell from https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#example-including-branches-and-tags the `on: push: branches: [master]` just evaluates against the git ref (available via) `github.ref` in the workflow. That will either be of the form `refs/head/<branch name>` or `refs/tags/<tag>`. So I (if I understand correctly the gh docu) `on: push: branches: [master]` is the same as `on: push` and then in the job testing for `github.ref == 'refs/heads/master`.

Comment (wvangeit) [reply .github/workflows/publish-python-client.yml:184]: Yeah, I think it definitely requires splitting the files in relevant parts. I think it anyway makes sense to have a separate workflow file that only runs on master.

Comment (bisgaard-itis) [reply .github/workflows/publish-python-client.yml:184]: Makes sense, thanks for the feedback. I will try splitting out the publishing of the package into a separate workflow tomorrow morning

File: scripts/generate_version.bash [hunk 3]
```diff
@@ -0,0 +1,17 @@
+#!/bin/bash
+
+# Generate version <semantic version of last tag(=release)>+<number of commits from last tagged commit>
```
Comment (pcrespov) [inline scripts/generate_version.bash:3]: THOUGHT:  using the number of commits from the last tag its a good idea but why not simply using the commit number i.e. short version of it. 

Moreover. Have a look at python's version identification scheme https://peps.python.org/pep-0440/#version-scheme
```
[N!]N(.N)*[{a|b|rc}N][.postN][.devN]
```
Chatty explains it very nicely:

1. `[N!]`: This part is typically used to represent the major version number. The "N" here represents a numeric digit. The "!" is optional and indicates that this part is "bang-prefixed," which suggests that this is an "epoch" version, and it's used to signify a major breaking change in the software that requires special attention.

2. `N(.N)*`: This part represents the minor version number. The "N" is a numeric digit, and the "(.N)*" means that there can be one or more additional digits separated by a dot, indicating sub-minor or patch versions. For example, it can be "2.1," "2.1.3," and so on.

3. `[{a|b|rc}N]`: This part is used to represent pre-release versions. The letters in square brackets indicate pre-release version types:
   - "a" for alpha
   - "b" for beta
   - "rc" for release candidate
   The "N" represents a numeric digit to indicate the specific version of the pre-release. For example, "a3" could represent alpha version 3.

4. `[.postN]`: This part represents a post-release version, typically used for bug fixes or minor updates after the initial release. The "N" here is a numeric digit indicating the specific post-release version.

5. `[.devN]`: This part can be used to indicate development versions or builds. The "N" represents a numeric digit to specify the specific development version or build number.

In summary, this version scheme provides a structured way to represent software versions with major, minor, pre-release, post-release, and development version components. It allows for clear version differentiation and helps users understand the nature of each version (e.g., whether it's a major release, a pre-release version, a post-release update, or a development build).

---

So after reading this, how would be adapted. IMO 
- the release tag could handle `[N!]N(.N)*`
- what about the rest??

File: clients/python/client/setup.py [hunk 4]
```diff
@@ -1,15 +1,12 @@
-import json
 from pathlib import Path
-from typing import Any, Dict
 
 from setuptools import find_packages, setup  # noqa: H301
 
-repo_root: Path = (Path(__file__) / "../../../..").resolve()
-
-config: Dict[str, Any] = json.loads((repo_root / "api/config.json").read_text())
+VERSION_FILE: Path = Path(__file__).parent / "VERSION"
```
Comment (pcrespov) [inline clients/python/client/setup.py:5]: Q: so this file does not exists in the repo but is generated in the CI, right? or we can create it using `make client/VERSION`

Comment (pcrespov) [reply clients/python/client/setup.py:5]: Q: did you add it in the `.gitignore`?

Comment (bisgaard-itis) [reply clients/python/client/setup.py:5]: Yes, I added a make recipe for this https://github.com/ITISFoundation/osparc-simcore-clients/blob/39b49dd9a9f33365959acb127b5338cefb5efd9d/clients/python/Makefile#L18. I also added it to `.gitignore`

File: .github/workflows/publish-python-client.yml (hunk 2 above)
Comment (bisgaard-itis) [reply .github/workflows/publish-python-client.yml:64]: @wvangeit I setup a new workflow for handling the publishing to Pypi.

File: scripts/generate_version.bash (hunk 3 above)
Comment (bisgaard-itis) [reply scripts/generate_version.bash:3]: Here are my thoughts on this: 

1. The reason for using the "distance" to the latest released/tagged commit is that the versions are linearly ordered. So it is easy to say if version A is older/newer than version B. That soothes my OCD and it is quite useful for testing because I can filter out tests by simply adding a `@pytest.skipif(Version(osparc.__version__)) <=Version("0.65.2+4")`. This gets more complicated if I use the commit hash directly. I can see the usefulness of having the commit hash accessible from within the package, but then I would add rather add it in some metadata, I think it is not so useful to have programatically available (whereas I think the version should be).
2. After reading your/chatty's summary I would suggest we something like `0.5..2.dev8`, where the `0.5.2` is the semantic version of the latest release and `dev8` indicates that this client version is 8 commits after the `0.5.2` release.

File: .github/workflows/publish-python-client.yml (hunk 2 above)
Comment (wvangeit) [reply .github/workflows/publish-python-client.yml:64]: Very good @bisgaard-itis

**Review by wvangeit — APPROVED** [review]

**Review by pcrespov — APPROVED** [review]
:tada: thx!
